RecoLocalCalo/EcalRecProducers/plugins/TimeComputationKernels.cu

```diff
@@ -0,0 +1,1143 @@
+#include <iostream>
+#include <limits>
+
+#include <cuda.h>
+
+#include "DataFormats/EcalDigi/interface/EcalDataFrame.h"
+#include "DataFormats/EcalRecHit/interface/EcalUncalibratedRecHit.h"
+#include "DataFormats/Math/interface/approx_exp.h"
+#include "DataFormats/Math/interface/approx_log.h"
+
+#include "Common.h"
+#include "TimeComputationKernels.h"
+#include "KernelHelpers.h"
+
+//#define DEBUG
+
+//#define ECAL_RECO_CUDA_DEBUG
+
+namespace ecal {
+  namespace multifit {
+
+    __device__ __forceinline__ bool use_sample(unsigned int sample_mask, unsigned int sample) {
+      return sample_mask & (0x1 << (EcalDataFrame::MAXSAMPLES - (sample + 1)));
+    }
+
+    __global__ void kernel_time_compute_nullhypot(SampleVector::Scalar const* sample_values,
+                                                  SampleVector::Scalar const* sample_value_errors,
+                                                  bool const* useless_sample_values,
+                                                  SampleVector::Scalar* chi2s,
+                                                  SampleVector::Scalar* sum0s,
+                                                  SampleVector::Scalar* sumAAs,
+                                                  const int nchannels) {
+      using ScalarType = SampleVector::Scalar;
+      constexpr int nsamples = EcalDataFrame::MAXSAMPLES;
+
+      // indices
+      int tx = threadIdx.x + blockDim.x * blockIdx.x;
+      int ltx = threadIdx.x;
+      int ch = tx / nsamples;
+      int nchannels_per_block = blockDim.x / nsamples;
+
+      // TODO: make sure that this branch plays nicely with __syncthreads inside
+      // can there be a deadlock even if the thread is inactive
```

Comment:
Should this TODO be addressed (the `__synchreads()` inside the if block scares me a bit)?

Reply:
@amassiro do you have a suggestion for this? Would having an `else` block with just three times `__syncthreads();` bring us back to safety? At the cost of being less efficient perhaps when?
Is the if just there as a protection or can it actually happen that `ch >= nchannels`?

Reply:
I think the branch can be changed from
```c++
if (ch < nchannels) {
  ...
}
```
to
```c++
if (ch >= nchannels)
  return;

 ...
```

With this change the `__syncthreads` should be fine, because threads that have exited the kernel shouldn't matter.

Reply:
Is there a way to rewrite the algorithms without adding sample 3 and removing it afterwards ?
Or, even better, to completely "unroll" the sums / subtractions to avoid the synchronisation ?

Reply:
This change you propose is actually already used in `RecoLocalCalo/EcalRecProducers/plugins/TimeComputationKernels.cu`. We can change all others as well.